### PR TITLE
VM Upgrade: T.S. '89

### PR DESCRIPTION
**Pull request by NullVoxPopuli**
image

Changelog: https://github.com/glimmerjs/glimmer-vm/releases/tag/v0.89.0

I had some issues with `this.attrs`, so I just remove the problematic tests, since `attrs` in general was supposed to be removed in v4, see: https://github.com/emberjs/ember.js/pull/20671 for that effort.

### CODE REVIEW

**Comment by chancancode** [comment]
Rebase after https://github.com/emberjs/ember.js/pull/20672 lands

File: packages/@ember/-internals/glimmer/lib/modifiers/action.ts [hunk 1]
```diff
@@ -200,6 +200,10 @@ class ActionModifierManager implements InternalModifierManager<ActionState, obje
     return new ActionState(element, owner, actionId, actionArgs, named, positional);
   }
 
+  getDebugInstance(state: ActionState): unknown {
+    return state;
+  }
```
Comment (chancancode) [inline packages/@ember/-internals/glimmer/lib/modifiers/action.ts:205]: I think we talked about this and we want to return `null`? We don't really want to leak the internal state bucket, and it's not really useful for it to show up on the inspector

File: packages/@ember/-internals/glimmer/lib/modifiers/internal.ts [hunk 2]
```diff
@@ -78,6 +78,10 @@ export class InternalModifierManager
     return null;
   }
 
+  getDebugInstance(state: InternalModifierState): unknown {
+    return state;
+  }
```
Comment (chancancode) [inline packages/@ember/-internals/glimmer/lib/modifiers/internal.ts:83]: I think just declare the class `abstract` and leave this out. We have to decide case-by-case what we want to expose here. And so far this is only used by `{{action}}` anyway

File: packages/@ember/-internals/glimmer/tests/integration/components/curly-components-test.js [hunk 3]
```diff
@@ -2492,7 +2492,7 @@ moduleFor(
           },
           value: null,
         }),
-        template: '<div id="inner-value">{{wrapper.content}}</div>',
+        template: '<div id="inner-value">{{this.wrapper.content}}</div>',
```
Comment (chancancode) [inline packages/@ember/-internals/glimmer/tests/integration/components/curly-components-test.js:2495]: Something is fishy/scary that these continued to work, seemingly without assertions/deprecations either. Hopefully the new VM code is erroring consistently now, but this means we may have implicit this code in the wild that was accidentally working

File: packages/@ember/-internals/glimmer/tests/integration/helpers/hash-test.js [hunk 4]
```diff
@@ -217,92 +217,5 @@ moduleFor(
...
-
-      runTask(() => {
-        expectDeprecation(() => {
-          set(fooBarInstance.hash, 'firstName', 'Godfrey');
-          set(fooBarInstance.hash, 'lastName', 'Chan');
-        }, /You set the '.*' property on a {{hash}} object/);
-      });
-
-      this.assertText('Godfrey Chan');
-    }
```
Comment (chancancode) [inline packages/@ember/-internals/glimmer/tests/integration/helpers/hash-test.js:306]: This went from deprecation to nothing (instead of an assertion). I don't think it's illegal per-se, I think we are technically allowed to do it, but that's when we tend to get ourselves into trouble – see the `{{#with}}` thing recently. Next time at least consider whether we can keep an assertion around for at least 1-2 LTS.

File: packages/ember-template-compiler/lib/plugins/assert-against-attrs.ts [hunk 5]
```diff
@@ -81,9 +81,8 @@ function isAttrs(node: AST.PathExpression, symbols: string[]) {
   }
 
   if (name === 'attrs') {
-    if (node.this === true) {
-      node.parts.shift();
-      node.original = node.original.slice(5);
+    if (node.head.type === 'ThisHead') {
+      node.tail.shift();
```
Comment (chancancode) [inline packages/ember-template-compiler/lib/plugins/assert-against-attrs.ts:85]: Rebase away

File: packages/ember-template-compiler/lib/plugins/transform-resolutions.ts [hunk 6]
```diff
@@ -149,7 +149,9 @@ export default function transformResolutions(env: EmberASTPluginEnvironment): AS
 }
 
 function isLocalVariable(node: AST.PathExpression, hasLocal: (k: string) => boolean): boolean {
-  return !node.this && node.parts.length === 1 && hasLocal(node.parts[0]!);
+  return (
+    !(node.head.type === 'ThisHead') && node.tail.length === 1 && hasLocal(node.head.original!)
```
Comment (chancancode) [inline packages/ember-template-compiler/lib/plugins/transform-resolutions.ts:153]: Why the ! on `.original`? It should always exist?

File: packages/ember-template-compiler/lib/plugins/transform-wrap-mount-and-outlet.ts [hunk 7]
```diff
@@ -43,8 +43,10 @@ export default function transformWrapMountAndOutlet(env: EmberASTPluginEnvironme
     name: 'transform-wrap-mount-and-outlet',
 
     visitor: {
-      Program: node,
+      Template: node,
       ElementNode: node,
+      BlockStatement: node,
```
Comment (chancancode) [inline packages/ember-template-compiler/lib/plugins/transform-wrap-mount-and-outlet.ts:48]: this is wrong (see below)

File: packages/ember-template-compiler/lib/plugins/utils.ts [hunk 8]
```diff
@@ -12,25 +12,41 @@ export function isStringLiteral(node: AST.Expression): node is AST.StringLiteral
...
+  return node;
+}
+
 export function trackLocals() {
   let locals = new Map();
 
   let node = {
-    enter(node: AST.Program | AST.Block | AST.ElementNode) {
-      for (let param of node.blockParams) {
+    enter(node: AST.Template | AST.Block | AST.ElementNode | AST.BlockStatement) {
```
Comment (chancancode) [inline packages/ember-template-compiler/lib/plugins/utils.ts:31]: You don't want `AST.BlockStatement` here. `BlockStatement` is the `{{#foo}}...{{/foo}}`, the block params is on the default/inverse blocks (which should get visited by the `Block` visitor). 

There _is_ a `params` on `BlockStatement` but it means a completely different thing, it is the positional argument for the invocation, which has a type of `AST.Expression[]` hence you question above.

On the other hand, `node.params` on `AST.Template | AST.Block | AST.ElementNode` should all be `AST.VarHead[]`.

File: packages/ember-template-compiler/lib/plugins/utils.ts [hunk 9]
```diff
@@ -12,25 +12,41 @@ export function isStringLiteral(node: AST.Expression): node is AST.StringLiteral
...
+      let params = 'params' in node ? node.params : node.blockParams;
```
Comment (chancancode) [inline packages/ember-template-compiler/lib/plugins/utils.ts:32]: All the nodes you checked here should have `params`. Why this check? Unless you intend for this `enter` hook to be more generic, in which case you need to type `node: AST.Node` and do a different check here, because there are other things with a `params` property that is unrelated.

File: packages/ember-template-compiler/lib/plugins/utils.ts [hunk 10]
```diff
@@ -12,25 +12,41 @@ export function isStringLiteral(node: AST.Expression): node is AST.StringLiteral
...
+      for (let param of params) {
+        let name = getLocalName(param);
```
Comment (chancancode) [inline packages/ember-template-compiler/lib/plugins/utils.ts:34]: Once you fix the types/guards `getLocalName` shouldn't need to exist

File: packages/ember-template-compiler/tests/plugins/assert-splattribute-expression-test.js [hunk 11]
```diff
@@ -1,48 +0,0 @@
-import { moduleFor, AbstractTestCase } from 'internal-test-helpers';
```
Comment (chancancode) [inline packages/ember-template-compiler/tests/plugins/assert-splattribute-expression-test.js:1]: I think we should keep this around and just change to use `assert.throws`. Generally we want to keep the coverage even if it duplicates what is already tested in the VM. The rule of thumb is if some day we decided to replace `glimmer-vm` with something else, we shouldn't loose coverage on anything that ember cares about.

File: types/publish.mjs [hunk 12]
```diff
@@ -109,8 +109,7 @@ async function main() {
 
   // The majority of those items should be excluded entirely, but in some cases
   // we still need to post-process them.
-  let excludes = remappedLocationExcludes
-    .concat(sideEffectExcludes);
+  let excludes = remappedLocationExcludes.concat(sideEffectExcludes);
```
Comment (chancancode) [inline types/publish.mjs:112]: This seems unrelated, it doesn't hurt but I'd revert it just to keep the diff minimal. I did notice we don't really have prettier set up here, which we can add in a different PR if we want.

File: packages/ember-template-compiler/tests/plugins/transform-component-invocation-test.js
Comment (chancancode) [inline packages/ember-template-compiler/tests/plugins/transform-component-invocation-test.js:1]: I don't know what this test file is trying to test anymore 😐

This probably corresponds to some kind of transform we did here to make some of these syntax work before they were implemented in the VM (this file probably predates the VM even existing), but we don't do that anymore, and the test doesn't actually do anything other than, checking these syntax compiles without error. 

Surely I hope that is already covered indirectly by any tests that actually verifies/exercises the feature. This is not that different from smoke testing `<p>hi</p>` and `{{this.foo}}` compiles without error, which, I suppose isn't inherently useless, but if we are going to do that we would want to add a more cases.

Not really your problem but I wonder if we should just delete this whole file.

File: packages/@ember/-internals/glimmer/lib/modifiers/internal.ts (hunk 2 above)
Comment (NullVoxPopuli) [reply packages/@ember/-internals/glimmer/lib/modifiers/internal.ts:83]: TS says its required, -- I'll return null by default tho

File: packages/@ember/-internals/glimmer/lib/modifiers/action.ts (hunk 1 above)
Comment (NullVoxPopuli) [reply packages/@ember/-internals/glimmer/lib/modifiers/action.ts:205]: right! updated

File: packages/ember-template-compiler/lib/plugins/transform-resolutions.ts (hunk 6 above)
Comment (NullVoxPopuli) [reply packages/ember-template-compiler/lib/plugins/transform-resolutions.ts:153]: yes, but nothing complained about it. It's been removed now.

File: packages/ember-template-compiler/lib/plugins/utils.ts (hunk 9 above)
Comment (NullVoxPopuli) [reply packages/ember-template-compiler/lib/plugins/utils.ts:32]: `params does not exist on type Template` - we can play with removing it I guess, but I feel like it was needed for something and I don't remember -- I'll see if I rediscover that tomorrow

File: packages/ember-template-compiler/lib/plugins/utils.ts (hunk 10 above)
Comment (NullVoxPopuli) [reply packages/ember-template-compiler/lib/plugins/utils.ts:34]: Still needs to exist because `param` can be of type string (from `blockParams`), which exist on `Template`

File: types/publish.mjs (hunk 12 above)
Comment (NullVoxPopuli) [reply types/publish.mjs:112]: yeah, it's annoying to have an editor that enables prettier when a `.prettierrc.js` is detected, but some files haven't been formatted with it. :(

Comment (NullVoxPopuli) [reply types/publish.mjs]: updated

File: packages/ember-template-compiler/lib/plugins/transform-wrap-mount-and-outlet.ts (hunk 7 above)
Comment (NullVoxPopuli) [reply packages/ember-template-compiler/lib/plugins/transform-wrap-mount-and-outlet.ts]: thanks! updated!

File: packages/ember-template-compiler/lib/plugins/assert-against-attrs.ts (hunk 5 above)
Comment (NullVoxPopuli) [reply packages/ember-template-compiler/lib/plugins/assert-against-attrs.ts:85]: other PR isn't merged yet :sweat_smile:

File: packages/ember-template-compiler/lib/plugins/utils.ts (hunk 9 above)
Comment (chancancode) [reply packages/ember-template-compiler/lib/plugins/utils.ts:32]: oh yeah, that’s right, cause we don’t have locs for the js scope locals 👍

File: packages/ember-template-compiler/lib/plugins/assert-against-attrs.ts (hunk 5 above)
Comment (chancancode) [reply packages/ember-template-compiler/lib/plugins/assert-against-attrs.ts:85]: Merged now!

Comment (NullVoxPopuli) [reply packages/ember-template-compiler/lib/plugins/assert-against-attrs.ts]: updated!

File: packages/ember-template-compiler/tests/plugins/assert-splattribute-expression-test.js (hunk 11 above)
Comment (NullVoxPopuli) [reply packages/ember-template-compiler/tests/plugins/assert-splattribute-expression-test.js:1]: `@glimmer/syntax` would still be used tho, even if we swap out the VM

Comment (NullVoxPopuli) [reply packages/ember-template-compiler/tests/plugins/assert-splattribute-expression-test.js:1]: (I'm intending on leaving this file deleted, partly because `@glimmer/syntax` is not glimmer-vm)

File: packages/@ember/-internals/glimmer/tests/integration/components/curly-components-test.js (hunk 3 above)
Comment (kategengler) [reply packages/@ember/-internals/glimmer/tests/integration/components/curly-components-test.js:2504]: I think `expectAssertion` below would have swallowed any other assertions, even if it matched on `backtrackingMessageFor`

File: packages/ember-template-compiler/tests/plugins/assert-splattribute-expression-test.js (hunk 11 above)
Comment (chancancode) [reply packages/ember-template-compiler/tests/plugins/assert-splattribute-expression-test.js:1]: Not necessarily, all those packages were invented out of thin air at the same time, it’s still an external (or internal, I guess?) dependency and ember needs to test its public api

Comment (NullVoxPopuli) [reply packages/ember-template-compiler/tests/plugins/assert-splattribute-expression-test.js:1]: It has it's own tests to tho. And we're planning on moving all that in to this repo

Comment (chancancode) [reply packages/ember-template-compiler/tests/plugins/assert-splattribute-expression-test.js:1]: Yes, that's what we have _always_ done. I didn't invent it just to slow down this particular PR. Often we have almost exactly the same test twice. See, for example: https://github.com/emberjs/ember.js/pull/17499

The tests are already there and it should be easy enough to change them to match (it's ok to not match the error messages very precisely, since the point is just to check that there are _some_ errors. 

Could we revisit how we do thing in general? Yes? But as you said you don't also want the creep the scope of the PR so unless this is a particularly important hill for you we can discuss that another time/day.

If it did turn out to be very challenging to keep the tests working we can of course revisit if it's worth the time. But as a matter of default, this is the status quo (and I don't think it's a pointless status quo/just useless formalities), and is what we should do.

Comment (NullVoxPopuli) [reply packages/ember-template-compiler/tests/plugins/assert-splattribute-expression-test.js:1]: The intent of this file has been restored, albeit, rewritten due to the nature of how the error was surfaced.

I removed the line checking, because, by the logic given, it's not relevant, and we want to make sure that syntax restrictions don't regress, due to not trusting `@glimmer/syntax` :sweat_smile: (which I don't agree with, but I can't care anymore :shrug: <3 )

File: packages/ember-template-compiler/tests/plugins/assert-splattribute-expression-test.js [hunk 13]
```diff
@@ -1,48 +1,27 @@
...
-      expectAssertion(() => {
-        compile('<div>{{...attributes}}</div>', {
-          moduleName: 'foo-bar',
-        });
-      }, this.expectedMessage(`'foo-bar' @ L1:C7`));
+    '@test {{...attributes}} is not valid path expression'(assert) {
+      assert.throws(() => compile('<div>{{...attributes}}</div>'), message);
     }
 
-    '@test {{...attributes}} is not valid path expression'() {
```
Comment (NullVoxPopuli) [inline packages/ember-template-compiler/tests/plugins/assert-splattribute-expression-test.js:25]: I deleted this test entirely because it was the exact same as the one above

File: packages/@ember/-internals/glimmer/lib/modifiers/internal.ts (hunk 2 above)
Comment (chancancode) [reply packages/@ember/-internals/glimmer/lib/modifiers/internal.ts:83]: Yeah, that's why I said *declare the class `abstract`*... when you _concretely_ implement/extend an abstract class, you are required to implement all the abstract members (which is the point of abstract classes). 

_However,_ you can also abstractly extend a class (`abstract class Foo extends AnotherAbstractClass`), in which case, you are allowed to leave out any abstract members, essentially deferring/propagating the requirement to your children

File: packages/@ember/-internals/glimmer/lib/modifiers/internal.ts [hunk 14]
```diff
@@ -78,6 +78,10 @@ export class InternalModifierManager
     return null;
   }
 
+  getDebugInstance(): unknown {
+    return null;
+  }
+
```
Comment (chancancode) [inline packages/@ember/-internals/glimmer/lib/modifiers/internal.ts:84]:
```suggestion
```

Comment (chancancode) [reply packages/@ember/-internals/glimmer/lib/modifiers/internal.ts:84]: Was hoping I could use this to commit the change directly from the UI, saving a roundtrip, but it wouldn't let me also add a diff for the `abstract class ...` because that line was unchanged 😐

Comment (NullVoxPopuli) [reply packages/@ember/-internals/glimmer/lib/modifiers/internal.ts]: this is a required method

```
Class 'InternalModifierManager' incorrectly implements interface 'InternalModifierManager<InternalModifierState, typeof InternalModifier>'.
     Property 'getDebugInstance' is missing in type 'InternalModifierManager' but required in type 'InternalModifierManager<InternalModifierState, typeof InternalModifier>'. [2420]
```

File: packages/@ember/-internals/glimmer/lib/modifiers/internal.ts [hunk 15]
```diff
@@ -78,9 +78,7 @@ export class InternalModifierManager
     return null;
   }
 
-  getDebugInstance(): unknown {
-    return null;
-  }
+  abstract getDebugInstance(state: InternalModifierState): unknown;
```
Comment (NullVoxPopuli) [inline packages/@ember/-internals/glimmer/lib/modifiers/internal.ts:81]: ah, this is what you meant

File: packages/@ember/-internals/glimmer/lib/modifiers/internal.ts (hunk 14 above)
Comment (NullVoxPopuli) [reply packages/@ember/-internals/glimmer/lib/modifiers/internal.ts]: nevermind: https://github.com/emberjs/ember.js/pull/20658/commits/a13c6275687d119a8de0dd9cb0f9755df2b8ba4d#r1544906209

File: packages/@ember/-internals/glimmer/lib/modifiers/internal.ts (hunk 15 above)
Comment (chancancode) [reply packages/@ember/-internals/glimmer/lib/modifiers/internal.ts:81]: Yeah, it’s a little bit of an edge case because we are implementing an interface. If we are extending an abstract super class you wouldn’t have to explicitly propagate it like this.

File: packages/ember-template-compiler/lib/plugins/index.ts [hunk 16]
```diff
@@ -36,8 +34,6 @@ export const STRICT_MODE_TRANSFORMS = Object.freeze(
     TransformActionSyntax,
     TransformEachInIntoEach,
     TransformInElement,
-    AssertSplattributeExpressions,
-    TransformEachTrackArray,
```
Comment (chancancode) [inline packages/ember-template-compiler/lib/plugins/index.ts:40]: This is revealing a real issue: https://github.com/emberjs/ember.js/pull/20673

We have to add it back, and deal it somehow on the glimmer-vm side. My initial thought is that glimmer-vm can assume everything that starts with a `-` is a keyword that ember will provide at runtime and not error on those (since it couldn't have been a JavaScript identifier), but maybe that will mask too many real errors when people are migrating existing templates (not uncommon to have helpers starting with dash in apps).

Maybe we'll have to add a `KeywordHead` subnode, or just make it something you are very unlikely to type, like `__GLIMMER_KEYWORD__` or something. Not sure.

But it is a real problem. If we don't deal with it, we would have broken `{{outlet}}`, `{{mount}}`, {{#each-in}}`, `{{#in-element}}` (perhaps among other things) for strict mode templates

**Review by chancancode — APPROVED** [review]
